src/wallet/wallet2.cpp

```diff
@@ -3084,29 +2938,41 @@ void wallet2::get_unconfirmed_payments(std::list<std::pair<crypto::hash,wallet2:
 //----------------------------------------------------------------------------------------------------
 void wallet2::rescan_spent()
 {
+  std::vector<bool> spent;
+  std::vector<bool> spent_in_chain;
+  std::vector<bool> spent_in_pool;
+
+  // make a list of key images for all our outputs
+
   // This is RPC call that can take a long time if there are many outputs,
   // so we call it several times, in stripes, so we don't time out spuriously
-  std::vector<int> spent_status;
-  spent_status.reserve(m_transfers.size());
   const size_t chunk_size = 1000;
   for (size_t start_offset = 0; start_offset < m_transfers.size(); start_offset += chunk_size)
   {
     const size_t n_outputs = std::min<size_t>(chunk_size, m_transfers.size() - start_offset);
-    MDEBUG("Calling is_key_image_spent on " << start_offset << " - " << (start_offset + n_outputs - 1) << ", out of " << m_transfers.size());
-    COMMAND_RPC_IS_KEY_IMAGE_SPENT::request req = AUTO_VAL_INIT(req);
-    COMMAND_RPC_IS_KEY_IMAGE_SPENT::response daemon_resp = AUTO_VAL_INIT(daemon_resp);
-    for (size_t n = start_offset; n < start_offset + n_outputs; ++n)
-      req.key_images.push_back(string_tools::pod_to_hex(m_transfers[n].m_key_image));
-    m_daemon_rpc_mutex.lock();
-    bool r = epee::net_utils::invoke_http_json("/is_key_image_spent", req, daemon_resp, m_http_client, rpc_timeout);
-    m_daemon_rpc_mutex.unlock();
-    THROW_WALLET_EXCEPTION_IF(!r, error::no_connection_to_daemon, "is_key_image_spent");
-    THROW_WALLET_EXCEPTION_IF(daemon_resp.status == CORE_RPC_STATUS_BUSY, error::daemon_busy, "is_key_image_spent");
-    THROW_WALLET_EXCEPTION_IF(daemon_resp.status != CORE_RPC_STATUS_OK, error::is_key_image_spent_error, daemon_resp.status);
-    THROW_WALLET_EXCEPTION_IF(daemon_resp.spent_status.size() != n_outputs, error::wallet_internal_error,
+
+    std::vector<crypto::key_image> key_images;
+    key_images.resize(n_outputs);
+    for (size_t i = start_offset; i < n_outputs; ++i)
+    {
+      const transfer_details& td = m_transfers[i];
+      key_images[i] = td.m_key_image;
+    }
+
+    std::vector<bool> spent_tmp;
+    std::vector<bool> spent_in_chain_tmp;
+    std::vector<bool> spent_in_pool_tmp;
+
+    boost::optional<std::string> r = m_daemon->keyImagesSpent(key_images, spent_tmp, spent_in_chain_tmp, spent_in_pool_tmp);
```

Comment:
Given start_offset is not passed, the filling of key_images seems wrong, since it can't tell where the "real" key images are.

Reply:
noted above, thanks.